lib/TH/THTensorApply.h

```diff
@@ -370,77 +501,84 @@
  * Tensor. But because we are guaranteed the subsequent data is contiguous in memory, we
  * can simply loop for sizeof(A) iterations and perform the operation, without having to
  * follow the order described by the strides of A.
+ *
+ * 3. We find the contiguous sections in a tensor and compress the counter array.
+ * For instance, if A is a 3x3x4x3 tensor, from which we get A' (3x3x3x3). Then, the
+ * original counter for A' is the same as A. But actually the first dimension and the
+ * second dimension can be collapsed into a single one.
```

Comment:
As an additional optimization, you could sort dimensions in order of decreasing stride. This would mean you'd get maximally contiguous sections. But even more importantly, it guarantees maximum data locality, which is very important for cache efficiency.

I know you already handle transposed tensors above with a special case, but for example you might have
```
t = torch.Tensor(3,3,4,3).narrow(2,0,3).transpose(1,3)
```